tests/test_expected_protocol.py

```diff
@@ -0,0 +1,123 @@
+                            ]) as instr:
+        assert instr.simple == 3.14
+        instr.simple = 4.5
+
+
+def test_error_checks():
+    """Test protocol for getting and setting with error checks."""
+    with expected_protocol(BasicTestInstrument,
+                           [('VOLT?', 3.14),
+                            ('SYST:ERR?', '0, 0'),
+                            ('VOLT 4.5 V', None),
+                            ('SYST:ERR?', '0, 0'),
+                            ]) as instr:
+        assert instr.with_error_checks == 3.14
+        instr.with_error_checks = 4.5
+
+
+def test_not_all_communication_used():
+    """Test whether unused communication raises an error."""
+    with raises(AssertionError, match="Unprocessed protocol definitions remain"):
+        with expected_protocol(BasicTestInstrument,
+                               [('VOLT?', 3.14),
+                                ('VOLT 4.5 V', None),
+                                ]) as instr:
+            assert instr.simple == 3.14
+
+
+def test_non_empty_write_buffer():
+    with raises(AssertionError, match="Non-empty write buffer"):
+        with expected_protocol(BasicTestInstrument,
+                               [('VOLT?', 3.14),
+                                ]) as instr:
+            instr.adapter.write_bytes(b"VOLT")
+            instr.adapter._index = 1
+
+
+def test_non_empty_read_buffer():
+    with raises(AssertionError, match="Non-empty read buffer"):
+        with expected_protocol(BasicTestInstrument,
+                               [('VOLT?', 3.14),
+                                ]) as instr:
+            instr.write("VOLT?")
+
+
+def test_preprocess_reply_on_values():
+    class InstrumentWithPreprocessValues(BasicTestInstrument):
+        """Workaround to get preprocess_reply working with protocol tests."""
+        def values(self, command, **kwargs):
+            return super().values(command, preprocess_reply=lambda v: v + "2345", **kwargs)
+
+    with expected_protocol(InstrumentWithPreprocessValues,
+                           [("VOLT?", "3.1")]) as instr:
+        assert instr.simple == 3.12345
+
+
+@pytest.mark.xfail
+def test_preprocess_reply_on_init():
+    # TODO: For now preprocess_reply at init level is nonfunctional because this
+    #  lives in the real adapter, should be fixed with #567
+    class InstrumentWithPreprocess(BasicTestInstrument):
+        def __init__(self, adapter, **kwargs):
+            super().__init__(adapter, preprocess_reply=lambda v: v + "2345", **kwargs)
+
+    with expected_protocol(InstrumentWithPreprocess,
+                           [("VOLT?", "3.1")]) as instr:
+        assert instr.simple == 3.12345
+
+
+# After completing expected_protocol tests, add tests elsewhere:
+# TODO: Add protocol tests for a simple instrument
+# TODO: Add protocol tests for an instrument with multiple channels
+# TODO: Add protocol tests for a frame-based instrument (recommend: TC038, TC038D)
```

Comment:
Done. We have yet to decide on the naming scheme: tc038d_test.py or test_tc038d.py. I started the documentation (PR coming), where it has to be included.